core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -911,14 +916,14 @@ private[spark] class TaskSetManager(
     logDebug("Checking for speculative tasks: minFinished = " + minFinishedForSpeculation)
     if (tasksSuccessful >= minFinishedForSpeculation && tasksSuccessful > 0) {
       val time = clock.getTimeMillis()
-      val durations = taskInfos.values.filter(_.successful).map(_.duration).toArray
-      Arrays.sort(durations)
+      val durations = successfulTasksSet.toArray.map(taskInfos(_).duration)
       val medianDuration = durations(min((0.5 * tasksSuccessful).round.toInt, durations.length - 1))
       val threshold = max(SPECULATION_MULTIPLIER * medianDuration, minTimeToSpeculation)
       // TODO: Threshold should also look at standard deviation of task durations and have a lower
       // bound based on that.
       logDebug("Task length threshold for speculation: " + threshold)
-      for ((tid, info) <- taskInfos) {
+      for (tid <- runningTasksSet) {
+        val info = taskInfos(tid)
```

Comment:
@kayousterhout 
Thanks a lot for your comments :)
I will keep this simple change in this pr. For time complexity improvement, I will make another pr and try add some measurements.